algosdk/abi.py

```diff
@@ -0,0 +1,570 @@
+            child_type=arg_type,
+            child_types=list(),
+            static_length=array_len,
+        )
+
+    @staticmethod
+    def make_address_type():
+        return Type(BaseType.Address)
+
+    @staticmethod
+    def make_dynamic_array_type(arg_type):
+        assert isinstance(arg_type, Type)
+        return Type(
+            BaseType.ArrayDynamic, child_type=arg_type, child_types=list()
+        )
+
+    @staticmethod
+    def make_string_type():
+        return Type(BaseType.String)
+
+    @staticmethod
+    def make_tuple_type(arg_types):
+        if len(arg_types) >= 2 ** 16:
+            raise error.ABITypeError(f"tuple args exceed 2^16")
+        assert isinstance(arg_types, list)
+        return Type(
+            BaseType.Tuple, child_types=arg_types, static_length=len(arg_types)
+        )
+
+    @staticmethod
+    def type_from_string(s):
+        if s.endswith("[]"):
+            array_arg_type = Type.type_from_string(s[:-2])
+            return Type.make_dynamic_array_type(array_arg_type)
+        elif s.endswith("]"):
+            static_array_regex = "^([a-z\d\[\](),]+)\[([1-9][\d]*)]$"
+            matches = re.search(static_array_regex, s)
+            try:
+                static_length = int(matches.group(2))
+                array_type = Type.type_from_string(matches.group(1))
+                return Type.make_static_array_type(array_type, static_length)
+            except:
+                raise error.ABITypeError(f"malformed static array string: {s}")
+        elif s.startswith("uint"):
+            try:
+                type_size = int(s[4:])
+                return Type.make_uint_type(type_size)
+            except:
+                raise error.ABITypeError(f"malformed uint string: {s}")
+        elif s == "byte":
+            return Type.make_byte_type()
+        elif s.startswith("ufixed"):
+            ufixed_regex = "^ufixed([1-9][\d]*)x([1-9][\d]*)$"
+            matches = re.search(ufixed_regex, s)
+            try:
+                bit_size = int(matches.group(1))
+                precision = int(matches.group(2))
+                return Type.make_ufixed_type(bit_size, precision)
+            except:
+                raise error.ABITypeError(f"malformed ufixed string: {s}")
+        elif s == "bool":
+            return Type.make_bool_type()
+        elif s == "address":
+            return Type.make_address_type()
+        elif s == "string":
+            return Type.make_string_type()
+        elif len(s) >= 2 and s[0] == "(" and s[-1] == ")":
+            # Recursively parse parentheses from a tuple string
+            tuples = Type.parse_tuple(s[1:-1])
+            tuple_list = []
+            for tup in tuples:
+                if isinstance(tup, str):
+                    tt = Type.type_from_string(str(tup))
+                    tuple_list.append(tt)
+                elif isinstance(tup, list):
+                    tts = list(map(lambda t_: Type.type_from_string(t_), tup))
```

Comment:
nit: is this the same as `tts = [Type.type_from_string(t_) for t_ in tup]`? If so, the list comprehension seems simpler to understand